### PR TITLE
[V4] Enhancements for the new sanitize processes

### PR DESCRIPTION
**Pull request by Convly**
Address issues raised in #11411 ([here](https://github.com/strapi/strapi/pull/11411#pullrequestreview-797865827))

- Fix typos
- Move the sanitize process for the eventHub to the eventHub domain
- Move the pipeAsync util from the sanitize utils to @strapi/utils
- Sanitize user information for email templates

### CODE REVIEW

File: packages/plugins/users-permissions/server/services/user.js [hunk 1]
```diff
@@ -121,22 +121,31 @@ module.exports = ({ strapi }) => ({
   async sendConfirmationEmail(user) {
     const userPermissionService = getService('users-permissions');
     const pluginStore = await strapi.store({ type: 'plugin', name: 'users-permissions' });
+    const userSchema = strapi.getModel('plugin::users-permissions.user');
 
     const settings = await pluginStore
       .get({ key: 'email' })
       .then(storeEmail => storeEmail['email_confirmation'].options);
 
+    // Sanitize the template's user information
```
Comment (alexandrebodin) [inline packages/plugins/users-permissions/server/services/user.js:130]: that's a lot of abstraction leaking here

File: packages/core/strapi/lib/services/event-hub.js [hunk 2]
```diff
@@ -5,7 +5,28 @@
...
+   * @param {object | object[]} entity
+   * @param {object} schema
+   * @returns {object}
+   */
+  sanitizeEntity(entity, schema) {
+    if (Array.isArray(entity)) {
+      return Promise.all(entity.map(entry => this.sanitizeEntity(entry, schema)));
+    }
+
+    const sanitizeFunction = pipeAsync(
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/services/event-hub.js:22]: Can this be sync or even the tranversEntity is async ?

Comment (alexandrebodin) [reply packages/core/strapi/lib/services/event-hub.js:22]: Same comment as later we are leaking quite a few things in there

Comment (Convly) [reply packages/core/strapi/lib/services/event-hub.js]: :+1:

File: packages/core/strapi/lib/services/entity-service/__tests__/entity-service.test.js [hunk 3]
```diff
@@ -21,7 +21,11 @@ describe('Entity service', () => {
       'Can decorate',
       async method => {
         const instance = createEntityService({
-          strapi: {},
+          strapi: {
+            eventHub: {
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/services/entity-service/__tests__/entity-service.test.js]: if you removed the function this shouldn't be necessary ?

File: packages/core/strapi/lib/services/entity-service/index.js [hunk 4]
```diff
@@ -60,7 +64,7 @@ const createDefaultImplementation = ({ strapi, db, eventHub, entityValidator })
 
   async emitEvent(uid, event, entity) {
     const model = strapi.getModel(uid);
-    const sanitizedEntity = await strapi.eventHub.sanitizeEntity(entity, model);
+    const sanitizedEntity = await sanitize.utils.defaultSanitizeOutput(model, entity);
```
Comment (alexandrebodin) [inline packages/core/strapi/lib/services/entity-service/index.js:97]: same question as above

**Review by alexandrebodin — COMMENTED** [review]
Overall looks good a few comments / questions

File: packages/core/strapi/lib/services/entity-service/__tests__/entity-service.test.js (hunk 3 above)
Comment (Convly) [reply packages/core/strapi/lib/services/entity-service/__tests__/entity-service.test.js]: Removed while you were reviewing :sweat_smile:

File: packages/core/admin/server/services/permission/permissions-manager/sanitize.js [hunk 5]
```diff
@@ -50,7 +50,7 @@ module.exports = ({ action, ability, model }) => {
       // Remove not allowed fields (RBAC)
       traverseEntity(allowedFields(permittedFields), { schema }),
       // Remove all fields of type 'password'
-      traverseEntity(removePassword, { schema })
+      sanitize.utils.sanitizePasswords(schema)
```
[the remark this answers is not in the document]
Comment (Convly) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:53]: Why not. We could also rename `utils` to `common` to better reflect the usage of what's inside? WDYT?

[the remark this answers is not in the document]
Comment (alexandrebodin) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:53]: like we have validate & validators we could have sanitize and sanitizers?

[the remark this answers is not in the document]
Comment (Convly) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js]: Alright, make sense, done

**Review by alexandrebodin — APPROVED** [review]
LGTM

**Pull request merged**